### PR TITLE
Fix issue with Ropper in macOS

### PR DESCRIPTION
**Pull request by Z6543**
Fix issue with Ropper in macOS -> AttributeError: 'Ropper' object has no attribute '__gatherGadgetsByEndings'

### CODE REVIEW

**Comment by epi052** [comment]
Howdy and thanks for the PR!

I took a look at the code, but I'm definitely not clear on how this fixes an attribute error.

Could you please enlighten me? I don't have a Mac to test the problem or the fix on.

**Comment by ZoltanBalazs** [comment]
Hi, 

the issue is with the buggy implementation of the multiprocessing module in macOS
I guess some modules start to load before others are loaded. 

Another solution was to turn of the multi-threaded functionality of Ropper by setting the option "multiprocessing" to false when invoking the Ropper service. 

Also, as this is a macOS specific bug, probably it is best to check the OS before invoking these 2 line of codes, and only use it in macOS

**Comment by epi052** [comment]
Ahhh, ok, that makes sense. Review incoming

File: find-gadgets.py [hunk 1]
```diff
@@ -276,7 +277,11 @@ def main(args):
     print_useful_regex(args.output, args.arch)
 
 
-if __name__ == "__main__":
+if __name__ == "__main__":    
+    #Fix issue with Ropper in macOS -> AttributeError: 'Ropper' object has no attribute '__gatherGadgetsByEndings'
+    multiprocessing.freeze_support()
```
Comment (epi052) [inline find-gadgets.py:282]: from the python docs: 

> Calling freeze_support() has no effect when invoked on any operating system other than Windows. In addition, if the module is being run normally by the Python interpreter on Windows (the program has not been frozen), then freeze_support() has no effect.

can remove this call, as far as i can tell

File: find-gadgets.py [hunk 2]
```diff
@@ -276,7 +277,11 @@ def main(args):
...
+    multiprocessing.set_start_method('fork')
```
Comment (epi052) [inline find-gadgets.py:283]: just like you suggested, let's gate this with an os check

```suggestion
    if platform.system() == "Darwin":
        multiprocessing.set_start_method('fork')
```

https://docs.python.org/3/library/platform.html#platform.system

File: find-gadgets.py [hunk 3]
```diff
@@ -6,6 +6,7 @@
 import tempfile
 import subprocess
 from pathlib import Path
+import multiprocessing
```
Comment (epi052) [inline find-gadgets.py:9]: we'll need to add another import for platform to satisfy os check

**Review by epi052 — CHANGES_REQUESTED** [review]

File: find-gadgets.py (hunk 1 above)
Comment (Z6543) [reply find-gadgets.py:282]: Yes, this can be removed

File: find-gadgets.py (hunk 2 above)
Comment (Z6543) [reply find-gadgets.py:283]: Let's not forget to add
import platform

**Review by Z6543 — COMMENTED** [review]
OK, just add import platform and ropper issues are fixed

But unfortunately this creates another problem with rp++, that the scripts parses rp++ results before rp++ returns ... Not sure what is the best way to solve this 

But this is a small bug compared to the original one.

**Comment by epi052** [comment]
I'm not entirely sure how to handle it, since the async stuff isn't in our actual code. You could try an Event/Condition/Lock to gate the access to the file reads until after the Gadgetizer has completed. But the code we own isn't async on its own, so I'm not sure how that would work.

If the above doesn't work, and simply disabling multiprocessing on mac is an option, i'd be ok with that.

**Comment by epi052** [comment]
@Z6543 sorry, i missed that you pushed some code to this awhile back; what's the state of this PR? Is it ready to merge?

**Comment by Z6543** [comment]
There is another challenge, the currently used rp++ version is not available as an "official" precompiled binary - and I am not sure how we should deal with that in this script. Or shall we move that problem to another issue/thread/pull request?

**Comment by epi052** [comment]
> There is another challenge, the currently used rp++ version is not available as an "official" precompiled binary - and I am not sure how we should deal with that in this script. Or shall we move that problem to another issue/thread/pull request?

yea, let's move that into a separate issue

**Comment by Z6543** [comment]
Once import platform is added, this should be good to go. Thx

**Comment by epi052** [comment]
not entirely sure what happened, but the current PR has no differences from main; mind re-adding the relevant changes?

**Comment by epi052** [comment]
fixed in #25

**Pull request closed without merging**